format.js

```diff
@@ -83,66 +130,63 @@ function makeEncryptionFormat() {
 
     if (!groupId) cb(new Error('Group id required'))
 
-    _keyringReady.onReady(() => {
-      cb(null, _keyring.group.get(groupId))
+    keyringReady.onReady(() => {
+      cb(null, keyring.group.get(groupId))
     })
   }
 
-  function addKeypair(keypair) {
-    _keyringReady.onReady(() => {
-      _keyring.dm.addFromSSBKeys(keypair)
-    })
+  function dmEncryptionKey(authorKeys, recp) {
+    const err = new Error('DM keys not supported for recipient ' + recp)
+    if (legacyMode) {
+      if (!keyring.dm.has(authorKeys.id, recp)) addDMPair(authorKeys, recp)
+      const dmKeys = keyring.dm.get(authorKeys.id, recp)
+      if (!dmKeys) throw err
+      return dmKeys
+    } else {
+      const theirRootId = recp
+      const myLeafId = authorKeys.id
+      const theirLeafId = keyring.dm.get(theirRootId, myLeafId)
```

Comment:
this looks a bit strange from an api point of view. Should we differentiate between this "getMirrorLeaf" function, and then getting the actual key (2 lines below)?

Reply:
Yeah, maybe should have a different getter function, just for the sake of naming. In keyring, the same leveldb instance is used to store these "triplets" as is for storing "pairs". Triplets help discover the leaf based on the root (because the root is the input recp), and pairs are just myLeaf & theirLeaf.

Storing triplets just helps us answer MF tree traversal questions (because the answers are stored) instead of having to actually crawl the tree structure every time we encrypt/decrypt. And storing pairs is about storing the actual DH keys.

Reply:
I named this `keyring.dm.triangulate`